### PR TITLE
Implement minimal update lifecycle

### LINKED ISSUE #60: Update resource lifecycle
Hi,

I wanted to ask if it was a deliberate decision to only allow create/delete, but no update lifecycle? We have a use case where it would be "more beautiful" if it could be seen as "in-place update" instead of a force replacement.

Would you be open for a PR on that topic?

### PR DESCRIPTION
**Pull request by christophkreutzer**
As discussed in #60 

@aidanmelen Would be interested in your feedback on this, thanks!

Fixes #60

### CODE REVIEW

File: snowsql/resource_exec.go [hunk 1]
```diff
@@ -136,7 +169,17 @@ func resourceExecRead(ctx context.Context, d *schema.ResourceData, m interface{}
 func resourceExecUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
 	var diags diag.Diagnostics
```
Comment (aidanmelen) [inline snowsql/resource_exec.go:171]: Let's also get the latest value of `name` just like we do in the create function. Then compare the new name with state. If we have a change we should update the Id and replace the object in state. For example:

```
resource "snowsql_exec" "dcl" {
  name = "my-name"

  create {
    statements = "CREATE USER IF NOT EXISTS user1;"
  }

  delete {
    statements = "DROP USER IF EXISTS user1;"
  }
}
```

and 

```
resource "snowsql_exec" "dcl" {
  name = "my-name-changed-should-cause-replacement"

  create {
    statements = "CREATE USER IF NOT EXISTS user1;"
  }

  delete {
    statements = "DROP USER IF EXISTS user1;"
  }
}
```

Comment (christophkreutzer) [reply snowsql/resource_exec.go:171]: I didn't test that, but doesn't name change force a replacement anyways as it is `ForceNew: true`?
https://github.com/aidanmelen/terraform-provider-snowsql/blob/main/snowsql/resource_exec.go#L67-L70

**Comment by aidanmelen** [comment]
start with this:

```hcl
resource "snowsql_exec" "user1" {
  name = "${local.name}_user1"

  create {
    statements = "CREATE USER IF NOT EXISTS ${local.name}_user1 PASSWORD='abc123' DEFAULT_ROLE = myrole DEFAULT_SECONDARY_ROLES = ('ALL') MUST_CHANGE_PASSWORD = TRUE;"
  }

  # update {
  #   statements = "ALTER USER IF EXISTS ${local.name}_user1 SET EMAIL = '${local.name}_user1@mail.com';"
  # }

  delete {
    statements = "DROP USER IF EXISTS ${local.name}_user1;"
  }
}
```

result on first run:

```bash
$ terraform apply --auto-approve
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # random_pet.server will be created
  + resource "random_pet" "server" {
      + id        = (known after apply)
      + length    = 2
      + prefix    = "SIMPLE_EXAMPLE"
      + separator = "_"
    }

  # snowsql_exec.user1 will be created
  + resource "snowsql_exec" "user1" {
      + id   = (known after apply)
      + name = (known after apply)

      + create {
          + number_of_statements = -1
          + statements           = (known after apply)
        }

      + delete {
          + number_of_statements = -1
          + statements           = (known after apply)
        }
    }

Plan: 2 to add, 0 to change, 0 to destroy.
random_pet.server: Creating...
random_pet.server: Creation complete after 0s [id=SIMPLE_EXAMPLE_tight_grubworm]
snowsql_exec.user1: Creating...
snowsql_exec.user1: Creation complete after 1s [id=SIMPLE_EXAMPLE_TIGHT_GRUBWORM_user1]

Apply complete! Resources: 2 added, 0 changed, 0 destroyed.
```

Then check the users email:

```sql
SHOW USERS LIKE '%SIMPLE_EXAMPLE_TIGHT_GRUBWORM_user1%';
-- name,created_on,login_name,display_name,first_name,last_name,email,mins_to_unlock,days_to_expiry,comment,disabled,must_change_password,snowflake_lock,default_warehouse,default_namespace,default_role,default_secondary_roles,ext_authn_duo,ext_authn_uid,mins_to_bypass_mfa,owner,last_success_login,expires_at_time,locked_until_time,has_password,has_rsa_public_key
-- SIMPLE_EXAMPLE_TIGHT_GRUBWORM_USER1,2023-02-10T07:52:29.434-08:00,SIMPLE_EXAMPLE_TIGHT_GRUBWORM_USER1,SIMPLE_EXAMPLE_TIGHT_GRUBWORM_USER1,,,,,,,false,true,false,,,MYROLE,"[""ALL""]",false,,,ACCOUNTADMIN,,,,true,false
```

Now update with email:

```hcl
resource "snowsql_exec" "user1" {
  name = "${local.name}_user1"

  create {
    statements = "CREATE USER IF NOT EXISTS ${local.name}_user1 PASSWORD='abc123' DEFAULT_ROLE = myrole DEFAULT_SECONDARY_ROLES = ('ALL') MUST_CHANGE_PASSWORD = TRUE;"
  }

  update {
    statements = "ALTER USER IF EXISTS ${local.name}_user1 SET EMAIL = '${local.name}_user1@mail.com';"
  }

  delete {
    statements = "DROP USER IF EXISTS ${local.name}_user1;"
  }
}
```

which results in the in-place update:

```bash
$ terraform apply --auto-approve
random_pet.server: Refreshing state... [id=SIMPLE_EXAMPLE_tight_grubworm]
snowsql_exec.user1: Refreshing state... [id=SIMPLE_EXAMPLE_TIGHT_GRUBWORM_user1]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # snowsql_exec.user1 will be updated in-place
  ~ resource "snowsql_exec" "user1" {
        id   = "SIMPLE_EXAMPLE_TIGHT_GRUBWORM_user1"
        name = "SIMPLE_EXAMPLE_TIGHT_GRUBWORM_user1"



      + update {
          + number_of_statements = -1
          + statements           = "ALTER USER IF EXISTS SIMPLE_EXAMPLE_TIGHT_GRUBWORM_user1 SET EMAIL = 'SIMPLE_EXAMPLE_TIGHT_GRUBWORM_user1@mail.com';"
        }
        # (2 unchanged blocks hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.

snowsql_exec.user1: Modifying... [id=SIMPLE_EXAMPLE_TIGHT_GRUBWORM_user1]
snowsql_exec.user1: Modifications complete after 1s [id=SIMPLE_EXAMPLE_TIGHT_GRUBWORM_user1]

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
```

Then check the users email again:

```sql
SHOW USERS LIKE '%SIMPLE_EXAMPLE_TIGHT_GRUBWORM_user1%';
-- name,created_on,login_name,display_name,first_name,last_name,email,mins_to_unlock,days_to_expiry,comment,disabled,must_change_password,snowflake_lock,default_warehouse,default_namespace,default_role,default_secondary_roles,ext_authn_duo,ext_authn_uid,mins_to_bypass_mfa,owner,last_success_login,expires_at_time,locked_until_time,has_password,has_rsa_public_key
-- SIMPLE_EXAMPLE_TIGHT_GRUBWORM_USER1,2023-02-10T07:52:29.434-08:00,SIMPLE_EXAMPLE_TIGHT_GRUBWORM_USER1,SIMPLE_EXAMPLE_TIGHT_GRUBWORM_USER1,,,SIMPLE_EXAMPLE_TIGHT_GRUBWORM_user1@mail.com,,,,false,true,false,,,MYROLE,"[""ALL""]",false,,,ACCOUNTADMIN,,,,true,false
```

apply idempontently without any changes:

```bash
$ terraform apply --auto-approve
random_pet.server: Refreshing state... [id=SIMPLE_EXAMPLE_tight_grubworm]
snowsql_exec.user1: Refreshing state... [id=SIMPLE_EXAMPLE_TIGHT_GRUBWORM_user1]

No changes. Your infrastructure matches the configuration.

Terraform has compared your real infrastructure against your configuration and found no differences, so no changes are needed.

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
```

Now change the name:

```

```hcl
resource "snowsql_exec" "user1" {
  name = "${local.name}_user1_changed"

  create {
    statements = "CREATE USER IF NOT EXISTS ${local.name}_user1 PASSWORD='abc123' DEFAULT_ROLE = myrole DEFAULT_SECONDARY_ROLES = ('ALL') MUST_CHANGE_PASSWORD = TRUE;"
  }

  update {
    statements = "ALTER USER IF EXISTS ${local.name}_user1 SET EMAIL = '${local.name}_user1@mail.com';"
  }

  delete {
    statements = "DROP USER IF EXISTS ${local.name}_user1;"
  }
}
```

which should result in the replacement of the user resource:

```bash
terraform apply --auto-approve
random_pet.server: Refreshing state... [id=SIMPLE_EXAMPLE_tight_grubworm]
snowsql_exec.user1: Refreshing state... [id=SIMPLE_EXAMPLE_TIGHT_GRUBWORM_user1]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
-/+ destroy and then create replacement

Terraform will perform the following actions:

  # snowsql_exec.user1 must be replaced
-/+ resource "snowsql_exec" "user1" {
      ~ id   = "SIMPLE_EXAMPLE_TIGHT_GRUBWORM_user1" -> (known after apply)
      ~ name = "SIMPLE_EXAMPLE_TIGHT_GRUBWORM_user1" -> "SIMPLE_EXAMPLE_TIGHT_GRUBWORM_user1_changed" # forces replacement



        # (3 unchanged blocks hidden)
    }

Plan: 1 to add, 0 to change, 1 to destroy.

snowsql_exec.user1: Destroying... [id=SIMPLE_EXAMPLE_TIGHT_GRUBWORM_user1]
snowsql_exec.user1: Destruction complete after 1s
snowsql_exec.user1: Creating...
snowsql_exec.user1: Creation complete after 1s [id=SIMPLE_EXAMPLE_TIGHT_GRUBWORM_user1_changed]

Apply complete! Resources: 1 added, 0 changed, 1 destroyed.
```

@christophkreutzer interestingly, the name change caused a replacement of the resource. I am going to invalidate my early comment.

File: snowsql/resource_exec.go [hunk 2]
```diff
@@ -136,7 +169,18 @@ func resourceExecRead(ctx context.Context, d *schema.ResourceData, m interface{}
...
 
-	// https://github.com/aidanmelen/terraform-provider-snowsql/issues/11
+	// TODO read before update
+
+	if d.HasChange("update.0.statements") {
```
Comment (aidanmelen) [inline snowsql/resource_exec.go:174]: Do you have any opinions about this change?

**Comment by christophkreutzer** [comment]
> @christophkreutzer interestingly, the name change caused a replacement of the resource. I am going to invalidate my early comment.

Yes, I think ForceNew is doing it's thing here :) But thanks for explicitly testing it!

File: snowsql/resource_exec.go (hunk 2 above)
Comment (christophkreutzer) [reply snowsql/resource_exec.go:174]: I'm fine with that! Much easier to read.

My reasoning was that update wouldn't even be called if it didn't change, but update what also be called if the delete lifecycle changed - and then, it's unnecessary to re-run the update script. So even cleaner as well.

**Comment by aidanmelen** [comment]
@christophkreutzer I also encountered an error when adding the update block and then removing  it during another apply. I think the most intuitive solution would be to do nothing, right?

**Comment by christophkreutzer** [comment]
Thank you very much for the swift review, additions and merging/releasing! :)

**Comment by christophkreutzer** [comment]
> @christophkreutzer I also encountered an error when adding the update block and then removing it during another apply. I think the most intuitive solution would be to do nothing, right?

You're right - would see the same behaviour (if update is removed, just remove it from the state and do nothing).
In that case, we probably need both checks in the update function: HasChange (did it actually change to the previous value) and checking the existence with GetOk.
https://github.com/hashicorp/terraform-plugin-sdk/blob/v2.20.0/helper/schema/resource_diff.go#L384

Another option would be to push that check further down to the parsing, but as create and delete are required, I see no advantage in doing so.

**Comment by aidanmelen** [comment]
I am working on adding generic read support. I can address this bug at that time.

**Comment by aidanmelen** [comment]
sorry, on second glance, the behavior on the main branch is good. It turns out that I was testing in a dev branch with the new read functionality.

Screen Shot 2023-02-11 at 7 31 39 AM

**Comment by aidanmelen** [comment]
@christophkreutzer please use the v1.1.1 release for the fix.

**Pull request merged**